### PR TITLE
Lake Formation Resource

### PR DESCRIPTION
**Pull request by gmazelier**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates #9700
Depends on #13250

**AWS Lake Formation support:** manages the data (Amazon S3 buckets and folders) that is being registered with AWS Lake Formation.

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
* **New Resource**: `aws_lakeformation_resource` (#13267)
```

Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSLakeFormationResource'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSLakeFormationResource -timeout 120m
=== RUN   TestAccAWSLakeFormationResource_basic
=== PAUSE TestAccAWSLakeFormationResource_basic
=== RUN   TestAccAWSLakeFormationResource_withRole
=== PAUSE TestAccAWSLakeFormationResource_withRole
=== RUN   TestAccAWSLakeFormationResource_update
=== PAUSE TestAccAWSLakeFormationResource_update
=== CONT  TestAccAWSLakeFormationResource_basic
=== CONT  TestAccAWSLakeFormationResource_update
=== CONT  TestAccAWSLakeFormationResource_withRole
--- PASS: TestAccAWSLakeFormationResource_basic (59.96s)
--- PASS: TestAccAWSLakeFormationResource_withRole (68.19s)
--- PASS: TestAccAWSLakeFormationResource_update (105.42s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	106.869s
```

### CODE REVIEW

File: aws/resource_aws_lakeformation_resource.go
Comment (ewbankkit) [inline aws/resource_aws_lakeformation_resource.go:24]: Maybe `validateArn` is better?

File: aws/resource_aws_lakeformation_resource.go
Comment (ewbankkit) [inline aws/resource_aws_lakeformation_resource.go:31]: Maybe `validateArn` is better?

File: aws/resource_aws_lakeformation_resource.go
Comment (gmazelier) [reply aws/resource_aws_lakeformation_resource.go:24]: Indeed, thanks for mentioning this function.

File: aws/resource_aws_lakeformation_resource.go
Comment (gmazelier) [reply aws/resource_aws_lakeformation_resource.go:31]: Will do.

**Comment by robertstettner** [comment]
This is really great! I started working on this as well, and managed to get the data sources for `data_aws_lakeformation_resource`, `data_aws_lakeformation_datalake_settings`, and `data_aws_lakeformation_permissions` to some good state. Maybe we should collaborate on this, and get the resource sorted for permissions? 😄

**Comment by gmazelier** [comment]
@robertstettner Yes it would be great! Good new for data sources support. In a draft pull request #13396 for permissions support. Work well for my first local tests, I need to improve acceptance tests but I will keep you posted in a few days to get your feedback.

**Comment by teamterraform** [comment]
## Notification of Recent and Upcoming Changes to Contributions

Thank you for this contribution! There have been a few recent development changes that affect this pull request. We apologize for the inconvenience, especially if there have been long review delays up until now. Please note that this is automated message from an unmonitored account. See the [FAQ](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/FAQ.md) for additional information on the maintainer team and review prioritization.

If you are unable to complete these updates, please leave a comment for the community and maintainers so someone can potentially continue the work. The maintainers will encourage other contributors to use the existing contribution as the base for additional changes as appropriate. Otherwise, contributions that do not receive updated code or comments from the original contributor may be closed in the future so the maintainers can focus on active items.

For the most up to date information about Terraform AWS Provider development, see the [Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md). Additional technical debt changes can be tracked with the [`technical-debt` label on issues](https://github.com/terraform-providers/terraform-provider-aws/issues?q=is%3Aopen+is%3Aissue+label%3Atechnical-debt).

As part of updating a pull request with these changes, the most current unit testing and linting will run. These may report issues that were not previously reported.

### Action Required: Terraform 0.12 Syntax

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/8950
Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14417

Version 3 and later of the Terraform AWS Provider, which all existing contributions would potentially be added, only supports Terraform 0.12 and later. Certain syntax elements of Terraform 0.11 and earlier show deprecation warnings during runs with Terraform 0.12. Documentation and test configurations, such as those including deprecated string interpolations (`some_attribute = "${aws_service_thing.example.id}"`) should be updated to the newer syntax (`some_attribute = aws_service_thing.example.id`). Contribution testing will automatically fail on older syntax in the near future. Please see the referenced issues for additional information.

### Action Required: Terraform Plugin SDK Version 2

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14551

The Terraform AWS Provider has been upgraded to the latest version of the Terraform Plugin SDK. Generally, most changes to contributions should only involve updating Go import paths in source code files. Please see the referenced issue for additional information.

### Action Required: Removal of website/aws.erb File

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14712

Any changes to the `website/aws.erb` file are no longer necessary and should be removed from this contribution to prevent merge issues in the near future when the file is removed from the repository. Please see the referenced issue for additional information.

### Upcoming Change of Git Branch Naming

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14292

Development environments will need their upstream Git branch updated from `master` to `main` in the near future. Please see the referenced issue for additional information and scheduling.

### Upcoming Change of GitHub Organization

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14715

This repository will be migrating from https://github.com/terraform-providers/terraform-provider-aws to https://github.com/hashicorp/terraform-provider-aws. No practitioner or developer action is anticipated and most GitHub functionality will automatically redirect to the new location. Go import paths including `terraform-providers` can remain for now. Please see the referenced issue for additional information and scheduling.

**Comment by YakDriver** [comment]
@gmazelier @ewbankkit I think this is in a pretty good place. I'm waiting to get the other PRs in shape, and add a few data sources before merging. If you have any feedback on this after the changes, please let me know!

File: aws/resource_aws_lakeformation_resource.go
Comment (ewbankkit) [inline aws/resource_aws_lakeformation_resource.go:93]:
```suggestion
		return fmt.Errorf("error reading Lake Formation Resource: %w", err)
```

File: aws/resource_aws_lakeformation_resource.go
Comment (ewbankkit) [inline aws/resource_aws_lakeformation_resource.go:119]:
```suggestion
		return fmt.Errorf("error updating Lake Formation Resource (%s): %w", d.Id(), err)
```

File: aws/resource_aws_lakeformation_resource.go
Comment (ewbankkit) [inline aws/resource_aws_lakeformation_resource.go:135]:
```suggestion
		return fmt.Errorf("error deregistering Lake Formation Resource (%s): %w", d.Id(), err)
```

File: aws/resource_aws_lakeformation_resource.go
Comment (ewbankkit) [inline aws/resource_aws_lakeformation_resource.go:108]: Maybe clearer to call `RegisterResource()` here directly rather than call `resourceAwsLakeFormationResourceCreate()`.
My head initially went to _Why is `role_arn` not `ForceNew: true`?_.

Comment (YakDriver) [reply aws/resource_aws_lakeformation_resource.go]: After more investigating, I agree with the original approach of @gmazelier. `ForceNew` is better. The `UpdateResource()` API call is quirky and only works for IAM role to IAM role updates. I added new acceptance tests, docs to show the limits.

**Comment by YakDriver** [comment]
Latest:

#### Acceptance tests, `us-gov-west-1`
```
--- PASS: TestAccAWSLakeFormationResource_disappears (28.22s)
--- PASS: TestAccAWSLakeFormationResource_serviceLinkedRole (28.24s)
--- PASS: TestAccAWSLakeFormationResource_basic (28.62s)
--- PASS: TestAccAWSLakeFormationResource_updateSLRToRole (50.82s)
--- PASS: TestAccAWSLakeFormationResource_updateRoleToRole (52.89s)
```

#### Acceptance tests, `us-west-2`

```
--- PASS: TestAccAWSLakeFormationResource_basic (27.39s)
--- PASS: TestAccAWSLakeFormationResource_disappears (28.05s)
--- PASS: TestAccAWSLakeFormationResource_serviceLinkedRole (29.05s)
--- PASS: TestAccAWSLakeFormationResource_updateRoleToRole (50.81s)
--- PASS: TestAccAWSLakeFormationResource_updateSLRToRole (50.88s)
```

**Comment by ewbankkit** [comment]
Verified acceptance tests (`us-west-2`, without service-linked role):

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSLakeFormationResource_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSLakeFormationResource_ -timeout 120m
=== RUN   TestAccAWSLakeFormationResource_basic
=== PAUSE TestAccAWSLakeFormationResource_basic
=== RUN   TestAccAWSLakeFormationResource_disappears
=== PAUSE TestAccAWSLakeFormationResource_disappears
=== RUN   TestAccAWSLakeFormationResource_serviceLinkedRole
=== PAUSE TestAccAWSLakeFormationResource_serviceLinkedRole
=== RUN   TestAccAWSLakeFormationResource_updateRoleToRole
=== PAUSE TestAccAWSLakeFormationResource_updateRoleToRole
=== RUN   TestAccAWSLakeFormationResource_updateSLRToRole
=== PAUSE TestAccAWSLakeFormationResource_updateSLRToRole
=== CONT  TestAccAWSLakeFormationResource_basic
=== CONT  TestAccAWSLakeFormationResource_updateRoleToRole
=== CONT  TestAccAWSLakeFormationResource_updateSLRToRole
=== CONT  TestAccAWSLakeFormationResource_serviceLinkedRole
=== CONT  TestAccAWSLakeFormationResource_disappears
=== CONT  TestAccAWSLakeFormationResource_updateSLRToRole
    provider_test.go:777: skipping tests; missing IAM service-linked role /aws-service-role/lakeformation.amazonaws.com. Please create the role and retry
--- SKIP: TestAccAWSLakeFormationResource_updateSLRToRole (0.91s)
=== CONT  TestAccAWSLakeFormationResource_serviceLinkedRole
    provider_test.go:777: skipping tests; missing IAM service-linked role /aws-service-role/lakeformation.amazonaws.com. Please create the role and retry
--- SKIP: TestAccAWSLakeFormationResource_serviceLinkedRole (0.92s)
--- PASS: TestAccAWSLakeFormationResource_basic (26.26s)
--- PASS: TestAccAWSLakeFormationResource_disappears (26.30s)
--- PASS: TestAccAWSLakeFormationResource_updateRoleToRole (47.26s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	47.385s
```

**Review by YakDriver — APPROVED** [review]

**Comment by gmazelier** [comment]
@YakDriver great!

**Pull request merged**
